[charts] Introduce keyboard navigation (#19155)
mui/mui-x: docs/data/charts/accessibility/accessibility.md

---
productId: x-charts
title: Charts - Accessibility
packageName: '@mui/x-charts'
---

# Accessibility

<p class="description">Learn how the Charts implement accessibility features and guidelines, including keyboard navigation that follows international standards.</p>

:::info
A common misconception about accessibility is to only consider blind people and the screen reader.
But there are other disability to consider, like:

- **Color blindness**, making it hard to distinguish different series, or low contrast elements.
- **Motion disability**, making it hard to open the tooltip on a given item.
- **Cognitive disability**, making it hard to focus your attention on some details.

> - **Vestibular dysfunction**, making you unconformable with animations.

Reviewer: ```suggestion
- **Vestibular dysfunction**, making you uncomfortable with animations.
```

This behavior is automated as much as possible, ensuring that the Date and Time Pickers are accessible in most cases.
A correct `aria-labelledby` value is assigned to the dialog component based on the following rules:

- Use `toolbar` id if the toolbar is visible;
- Use the id of the input label if the toolbar is hidden;

:::info
Make sure to provide an `aria-labelledby` prop to `popper` and/or `mobilePaper` slots in case you are using Date and Time Pickers component with **hidden toolbar** and **without a label**.
::: -->

## Animation

Most of the charts have an animation when rendering or when data update.
For users with vestibular motion disorders those animations can be problematic.
By default we skip animation based on the [`prefers-reduced-motion`](https://developer.mozilla.org/en-US/docs/Web/CSS/@media/prefers-reduced-motion) media feature.

<!--
## Screen reader compatibility

Date and Time Pickers use ARIA roles and robust focus management across the interactive elements to convey the necessary information to users, being optimized for use with assistive technologies.
-->

## Keyboard support

> Set `enableKeyboardNavigation` to `true` to enable the keyboard navigation on your charts.

Reviewer: Should we add a demo?

---
productId: x-charts
title: Charts - Accessibility
packageName: '@mui/x-charts'
---

# Accessibility

> <p class="description">Learn how the Charts implement accessibility features and guidelines, including keyboard navigation that follows international standards.</p>

Reviewer 2: ```suggestion
<p class="description">Learn how Charts implement accessibility features and guidelines, including keyboard navigation that follows international standards.</p>
```

Some charts have animations when rendering or when data updates.
For users with vestibular motion disorders those animations can be problematic.
By default animations are toggled based on the [`prefers-reduced-motion`](https://developer.mozilla.org/en-US/docs/Web/CSS/@media/prefers-reduced-motion) media feature.

<!--
## Screen reader compatibility

Date and Time Pickers use ARIA roles and robust focus management across the interactive elements to convey the necessary information to users, being optimized for use with assistive technologies.
-->

## Keyboard support

Set `enableKeyboardNavigation` to `true` to enable the keyboard navigation on your charts.
You can also enable it globally using [theme default props](/material-ui/customization/theme-components/#theme-default-props)

```js
components: {
  MuiChartDataProvider: {
    defaultProps: {
      enableKeyboardNavigation: true
    },
  },
}
```

{{"demo": "KeyboardNavigation.js"}}

This feature is currently supported by line, bar, pie, scatter, and sparkline charts.

This makes the SVG component focusable thanks to [`tabIndex`](https://developer.mozilla.org/en-US/docs/Web/HTML/Reference/Global_attributes/tabindex).

> When focused, the charts highlight a value item that can be modified with arrow navigation.

Reviewer 2: ```suggestion
When focused, the chart highlights a value item that can be modified with arrow navigation.
```
